src/main/java/com/amazon/opendistroforelasticsearch/search/asynchronous/service/AsynchronousSearchService.java

```diff
@@ -312,22 +313,17 @@ public void freeContext(String id, AsynchronousSearchContextId asynchronousSearc
         }
     }
 
-    private void cancelTask(AsynchronousSearchActiveContext asynchronousSearchContext, String reason, Runnable runnable) {
-        if (asynchronousSearchContext.getTask() != null && asynchronousSearchContext.getTask().isCancelled() == false) {
-            CancelTasksRequest cancelTasksRequest = new CancelTasksRequest()
-                    .setTaskId(new TaskId(clusterService.localNode().getId(), asynchronousSearchContext.getTask().getId()))
-                    .setReason(reason);
-            client.admin().cluster().cancelTasks(cancelTasksRequest, runAfter(wrap(cancelTasksResponse ->
-                            logger.debug("Successfully cancelled tasks [{}] with asynchronous search [{}] with response [{}]",
-                                    asynchronousSearchContext.getTask(), asynchronousSearchContext.getAsynchronousSearchId(),
-                                    cancelTasksResponse),
-                    e -> logger.error(() -> new ParameterizedMessage("Failed to cancel task [{}] with asynchronous search [{}]" +
-                            " with exception", asynchronousSearchContext.getTask(), asynchronousSearchContext.getAsynchronousSearchId()),
-                            e)),
-                    runnable));
-        } else {
-            runnable.run();
-        }
+    private void cancelTask(AsynchronousSearchActiveContext asynchronousSearchContext, String reason,
+                            ActionListener<CancelTasksResponse> listener) {
+        CancelTasksRequest cancelTasksRequest = new CancelTasksRequest()
+                .setTaskId(new TaskId(clusterService.localNode().getId(), asynchronousSearchContext.getTask().getId()))
+                .setReason(reason);
+        client.admin().cluster().cancelTasks(cancelTasksRequest, listener);
+
+    }
+    private boolean shouldCancel(AsynchronousSearchActiveContext asynchronousSearchContext) {
+        return asynchronousSearchContext.getTask() != null && asynchronousSearchContext.getTask().isCancelled() == false
+                && asynchronousSearchContext.isCompleted() ==  false;
```

Comment:
Remove ` asynchronousSearchContext.isCompleted() ==  false`

Reply:
retaining this.